File: Prototypes/Python/grasp.py
import random

class Item:
    def __init__(self, weight, profit, index):
        self.weight = weight
        self.profit = profit
        self.ratio = profit / weight
        self.index = index

class BlockingPair:
    def __init__(self, item1, item2):
        self.item1 = item1
        self.item2 = item2

# Calcula (Wi / Pi) para cada item
def calculate_ratios(items):
    for item in items:
        item.ratio = item.profit / item.weight

# Ordena os itens em ordem decrescente pela razão (elemento, razão)
def construct_LCR(items, pairs, alpha, selected_indices):
    items.sort(key=lambda x: x.ratio, reverse=True)  # Ordena os itens em ordem decrescente pela razão (elemento, razão)
    
    min_ratio = items[-1].ratio  # Calcula a razão mínima e máxima (extremidades da lista em 2)
    max_ratio = items[0].ratio
    threshold = min_ratio + alpha * (max_ratio - min_ratio)

    LCR = []
    for item in items:
        # Para cada elemento da lista 2, adiciono se:
        if min_ratio <= item.ratio <= threshold and not is_blocked(item.index, pairs, selected_indices):  # b) se o par do elemento não estiver na lista
            LCR.append(item)
    
    return LCR

def is_blocked(item_index, pairs, selected_indices):
    for pair in pairs:
        if pair.item1 == item_index and pair.item2 in selected_indices:
            return True
        if pair.item2 == item_index and pair.item1 in selected_indices:
            return True
    return False

# GRASP: os pares bloqueadores serão uma lista onde o índice indica o índice dos elementos do outro par.
def GRASP(items, capacity, pairs, alpha):
    selected_indices = []
    current_weight = 0
    solution = []

    # 1- Cria o conjunto S_ com todos os elementos
    # 2- Cria o conjunto S inicialmente vazio com os elementos que entrarão na mochila
    # 3- Cria variável W com o peso da mochila (inicialmente 0)
    while items and current_weight < capacity:  # 4- Enquanto |S_| != 0 e W < C:
        LCR = construct_LCR(items, pairs, alpha, selected_indices)  # 4.1 Construo a LCR passando a S_, os pares bloqueados e A_
        
        if not LCR:
            break

        selected_item = random.choice(LCR)  # 4.2 Escolho elementos aleatórios da LCR e retiro de S_
        
        if current_weight + selected_item.weight <= capacity:  # 4.3 se W+Wi <= C:
            solution.append(selected_item)  # 4.3.1 Adiciono (Wi, Pi) a S
            current_weight += selected_item.weight  # 4.3.2 W+= Wi
            selected_indices.append(selected_item.index)

        items.remove(selected_item)

    print("Selected items:")
    for item in solution:
        print(f"Item {item.index}: Weight = {item.weight}, Profit = {item.profit}")
    print(f"Total weight: {current_weight}")

if __name__ == "__main__":
    random.seed()

    # Definir itens e pares bloqueadores aqui
    items = [Item(2, 10, 0), Item(3, 14, 1), Item(4, 16, 2), Item(5, 30, 3)]
    pairs = [BlockingPair(0, 2), BlockingPair(1, 3)]
    capacity = 7
    alpha = 0.3

    GRASP(items, capacity, pairs, alpha)

File: Prototypes/C/grasp.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

typedef struct {
    int weight;
    int profit;
    double ratio;
    int index;
} Item;

typedef struct {
    int item1;
    int item2;
} BlockingPair;

// Calcula (Wi / Pi) para cada item
void calculate_ratios(Item items[], int n) {
    for (int i = 0; i < n; i++) {
        items[i].ratio = (double)items[i].profit / items[i].weight;
        items[i].index = i;
    }
}

// Ordena os itens em ordem decrescente pela razão (elemento, razão)
int compare_items(const void *a, const void *b) {
    double ratioA = ((Item *)a)->ratio;
    double ratioB = ((Item *)b)->ratio;
    return (ratioB - ratioA) > 0 ? 1 : -1;
}

// Verifica se o item está bloqueado pelos pares selecionados
int is_blocked(int itemIndex, BlockingPair pairs[], int numPairs, int selectedIndices[], int numSelected) {
    for (int i = 0; i < numPairs; i++) {
        if (pairs[i].item1 == itemIndex) {
            for (int j = 0; j < numSelected; j++) {
                if (pairs[i].item2 == selectedIndices[j]) return 1;
            }
        }
        if (pairs[i].item2 == itemIndex) {
            for (int j = 0; j < numSelected; j++) {
                if (pairs[i].item1 == selectedIndices[j]) return 1;
            }
        }
    }
    return 0;
}

// Construo a LCR da seguinte forma:
void construct_LCR(Item items[], int n, BlockingPair pairs[], int numPairs, double alpha, Item LCR[], int *lcrSize, int selectedIndices[], int numSelected) {
    qsort(items, n, sizeof(Item), compare_items); // Ordena os itens em ordem decrescente pela razão (elemento, razão)
    
    double minRatio = items[n-1].ratio; // Calcula a razão mínima e máxima (extremidades da lista em 2)
    double maxRatio = items[0].ratio;
    double threshold = minRatio + alpha * (maxRatio - minRatio);

    *lcrSize = 0;
    for (int i = 0; i < n; i++) {
        // Para cada elemento da lista 2, adiciono se:
        if (items[i].ratio >= minRatio && items[i].ratio <= threshold && 
            !is_blocked(items[i].index, pairs, numPairs, selectedIndices, numSelected)) { // b) se o par do elemento não estiver na lista
            LCR[(*lcrSize)++] = items[i];
        }
    }
}

// GRASP: os pares bloqueadores serão uma lista onde o índice indica o índice dos elementos do outro par.
void GRASP(Item items[], int n, int capacity, BlockingPair pairs[], int numPairs, double alpha) {
    Item *LCR = (Item *)malloc(n * sizeof(Item));
    int *selectedIndices = (int *)malloc(n * sizeof(int));
    int lcrSize, numSelected = 0;
    int currentWeight = 0;
    Item *solution = (Item *)malloc(n * sizeof(Item));
    int solutionSize = 0;
    
    // 1- Cria o conjunto S_ com todos os elementos
    // 2- Cria o conjunto S inicialmente vazio com os elementos que entrarão na mochila
    // 3- Cria variável W com o peso da mochila (inicialmente 0)
    while (n > 0 && currentWeight < capacity) { // 4- Enquanto |S_| != 0 e W < C:
        construct_LCR(items, n, pairs, numPairs, alpha, LCR, &lcrSize, selectedIndices, numSelected); // 4.1 Construo a LCR passando a S_, os pares bloqueados e A_
        
        if (lcrSize == 0) break;

        int randomIndex = rand() % lcrSize; // 4.2 Escolho elementos aleatórios da LCR e retiro de S_
        Item selectedItem = LCR[randomIndex];
        
        if (currentWeight + selectedItem.weight <= capacity) { // 4.3 se W+Wi <= C:
            solution[solutionSize++] = selectedItem; // 4.3.1 Adiciono (Wi, Pi) a S
            currentWeight += selectedItem.weight; // 4.3.2 W+= Wi
            selectedIndices[numSelected++] = selectedItem.index;
        }

        int removeIndex = -1;
        for (int i = 0; i < n; i++) {
            if (items[i].index == selectedItem.index) {
                removeIndex = i;
                break;
            }
        }
        
        if (removeIndex != -1) {
            for (int i = removeIndex; i < n - 1; i++) {
                items[i] = items[i + 1];
            }
            n--;
        }
    }

    printf("Selected items:\n");
    for (int i = 0; i < solutionSize; i++) {
        printf("Item %d: Weight = %d, Profit = %d\n", solution[i].index, solution[i].weight, solution[i].profit);
    }
    printf("Total weight: %d\n", currentWeight);
    
    free(LCR);
    free(selectedIndices);
    free(solution);
}

int main() {
    srand(time(NULL));

    // Definir itens e pares bloqueadores aqui
    Item items[] = {{2, 10}, {3, 14}, {4, 16}, {5, 30}};
    int numItems = sizeof(items) / sizeof(items[0]);
    BlockingPair pairs[] = {{0, 2}, {1, 3}};
    int numPairs = sizeof(pairs) / sizeof(pairs[0]);
    int capacity = 7;
    double alpha = 0.3;

    calculate_ratios(items, numItems);
    GRASP(items, numItems, capacity, pairs, numPairs, alpha);

    return 0;
}
